src/main/java/core/basesyntax/MyLinkedList.java

```
@@ -1,47 +1,139 @@
+        Node<T> node;
+        if (findNode.equals(first)) {
+            node = new Node<>(null, value, first);
+            first.prev = node;
+            first = node;
+        } else {
+            node = new Node<>(findNode.prev, value, findNode);
+            findNode.prev.next = node;
+            findNode.prev = node;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return findIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> findNode = findIndex(index);
+        T deletedNode = findNode.item;
+        findNode.item = value;
+        return deletedNode;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> node = findIndex(index);
+        unlink(node);
+        return node.item;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> node = first;
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(object, node.item)) {
+                unlink(node);
+                return true;
+            }
+            node = node.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void unlink(Node<T> node) {
+        if (node.prev == null) {
+            first = node.next;
+        } else if (node.next == null) {
+            last = node.prev;
+        } else {
+            node.prev.next = node.next;
+            node.next.prev = node.prev;
+        }
+        size--;
+    }
+
+    private boolean checkIndex(int index) {
+        return index >= 0 && index <= size;
```

we can throw exception in this method in case index is invalid